torch/compiler/__init__.py

```diff
@@ -216,18 +217,69 @@ def assume_constant_result(fn):
 
 def disable(fn=None, recursive=True):
     """
-    This function provides both a decorator and a context manager to disable compilation on a function
+    This function provides a decorator manager to disable compilation on a function
     It also provides the option of recursively disabling called functions
 
+    .. note::
+        Interaction between `compile`, `disable`, and `enable`
+
+        `compile` is is a "marker" that Dynamo should attempt to compile the function
+        and its nested calls.
+
+        `disable` is higher-priority - it signifies that a function (and its nested
+        calls in the case recursive=True) should not be compiled.
+
+        In particular, `disable` overrides `compile` - if you want to re-enable compilation,
+        use `enable`. `disable` and `enable` have the same priority.
+
+        e.g.
+
+            @enable
+            def a(x):
+                ...
+
+            @disable
+            def b(x):
+                a(x)
+                ...
+
+            @compile
+            def c(x):
+                b(x)
+                ...
+
+            @disable
+            def d(x):
+                c(x)
+                ...
+
+        Calling `a`  will result in no compilation.
+        Calling `b` will result in no compilation
+        Calling `c` will result in `c` and `a` being compiled.
+            A graph break will occur when `b` is called.
+        Calling `d` will result in `a` being compiled.
```

Comment:
Maybe we should also talk a little bit about disable recursive=False.  Non-recursive disable is equivalent to immediately "re-enabling" PT2 on all the subcalls right? (Is this what actually happens?)

Reply:
skip/recursive=False is implemented by applying `SKIP_CODE` to the current frame and marking the function with `_torchdynamo_disable`, which causes graph breaks if we attempt to inline. It does not touch `eval_frame_callback_enabled`.

There are some behaviors of skip/recursive=False that are not immediately obvious:
- subcalls are not actively re-enabled. If we are in a recursive=True disable region, subcalls continue to be disabled.
- nested compile/enable/disable decorators are skipped.

I think the simplest solution is to declare recursive=False as even higher priority than recursive=True/enable. Because recursive=False only applies to one frame, there is no need to "re-enable".